ocean_provider/routes/decrypt.py

```diff
@@ -0,0 +1,172 @@
+
+setup_logging()
+provider_wallet = get_provider_wallet()
+requests_session = get_requests_session()
+requests_session.mount("file://", LocalFileAdapter())
+
+logger = logging.getLogger(__name__)
+
+
+@services.route("/decrypt", methods=["POST"])
+@validate(DecryptRequest)
+def decrypt():
+    data = get_request_data(request)
+    logger.info(f"decrypt endpoint called. {data}")
+
+    try:
+        return _decrypt(
+            decrypter_address=data.get("decrypterAddress"),
+            chain_id=data.get("chainId"),
+            transaction_id=data.get("transactionId"),
+            data_nft_address=data.get("dataNftAddress"),
+            encrypted_document=data.get("encryptedDocument"),
+            flags=data.get("flags"),
+            document_hash=data.get("documentHash"),
+        )
+    except Exception as e:
+        return service_unavailable(e, data, logger)
+
+
+def _decrypt(
+    decrypter_address: HexStr,
+    chain_id: int,
+    transaction_id: Optional[HexStr],
+    data_nft_address: Optional[HexStr],
+    encrypted_document: Optional[HexStr],
+    flags: Optional[int],
+    document_hash: Optional[HexStr],
+) -> Response:
+    increment_nonce(decrypter_address)
+
+    # Check if given chain_id matches Provider's chain_id
+    web3 = get_web3()
+    if web3.eth.chain_id != chain_id:
+        return error_response(f"Unsupported chain ID", 400)
+
+    # Check if decrypter is authorized
+    authorized_decrypters = get_config().authorized_decrypters
+    logger.info(f"authorized_decrypters = {authorized_decrypters}")
+    if authorized_decrypters and decrypter_address not in authorized_decrypters:
+        return error_response(f"Decrypter not authorized", 403)
+
+    if transaction_id:
+        try:
+            tx_receipt = web3.eth.get_transaction_receipt(transaction_id)
+            logs = get_metadata_logs_from_tx_receipt(web3, tx_receipt)
+            logger.info(f"transaction_id = {transaction_id}, logs = {logs}")
+            if len(logs) > 1:
+                logger.warning(
+                    "More than 1 MetadataCreated/MetadataUpdated event detected. "
+                    "Using the event at index 0."
+                )
+            log = logs[0]
+            data_nft_address = log.address
+            # Interpret "data" as utf-8 encoded bytes
+            encrypted_document = log.args["data"]
+            # Interpret "flags" as array of bytes length 1
+            flags = log.args["flags"]
+            # Interpret metaDataHash" as utf-8 encoded bytes
+            document_hash = log.args["metaDataHash"]
+            logger.info(
+                f"data_nft_address = {data_nft_address}, "
+                f"encrypted_document = {encrypted_document}, "
+                f"flags = {flags}, "
+                f"document_hash = {document_hash}"
+            )
+        except Exception:
+            response = error_response(f"Failed to get metadata logs.", 400)
+            logger.error(f"{traceback.format_exc()}")
```

Comment:
let's move the logger before the error response and remove `response =` to return the error response directly, for concision

Reply:
Done.  Thanks for the comment!